File: goblinframework-core/src/main/java/org/goblinframework/core/event/GoblinEventFuture.java
package org.goblinframework.core.event;

import org.goblinframework.api.concurrent.GoblinFuture;
import org.goblinframework.api.function.Block0;
import org.jetbrains.annotations.NotNull;

/**
 * Goblin event future abstraction.
 *
 * @author Xiaohai Zhang
 * @since Dec 3, 2019
 */
public interface GoblinEventFuture extends GoblinFuture<GoblinEventContext> {

  /**
   * Callback the specified action in case of event future discarded.
   */
  void addDiscardListener(@NotNull Block0 action);
}

File: goblinframework-core/src/main/java/org/goblinframework/core/event/GoblinEventFuture.kt
package org.goblinframework.core.event

import org.goblinframework.api.concurrent.GoblinFuture
import org.goblinframework.core.concurrent.GoblinFutureImpl
import org.goblinframework.core.event.exception.EventBusException

class GoblinEventFuture : GoblinFutureImpl<GoblinEventContext>() {

  override fun complete(result: GoblinEventContext?): GoblinFuture<GoblinEventContext> {
    var error: EventBusException? = null
    (result as? GoblinEventContextImpl)?.run {
      error = this.throwExceptionIfNecessary()
    }
    return super.complete(result, error)
  }

  override fun complete(result: GoblinEventContext?, cause: Throwable?): GoblinFuture<GoblinEventContext> {
    throw UnsupportedOperationException()
  }

}
